services/src/test/java/com/chatbot/services/PubSubControllerTests.java

```diff
@@ -0,0 +1,123 @@
+        MockMvcRequestBuilders.post(uri).contentType(MediaType.APPLICATION_JSON_VALUE);
+
+  }
+
+  @Test
+  public void onEvent_suggestCategoryChange() throws Exception {
+    final JsonNode message = mapper
+        .readTree(AsyncServiceTests.class.getResourceAsStream("/pubSubSuggestCategoryChange.json"));
+    final MvcResult mvcResult = mockMvc.perform(mockRequestBuilder
+        .content(message.toString())).andReturn();       
+    assertEquals(200, mvcResult.getResponse().getStatus());
+    ArgumentCaptor<TriggerEventNotification> notification =
+        ArgumentCaptor.forClass(TriggerEventNotification.class);
+    verify(asyncService).triggerEventHandler(notification.capture());
+    assertEquals(TriggerEventNotification.Event.SUGGEST_CATEGORY_CHANGE,
+        notification.getValue().getEvent(), "Error parsing event");
+    assertEquals("cafe",
+        notification.getValue().getEventParams().getFieldsMap()
+        .get("suggestedCategory").getStringValue(),
+        "Error parsing event params");
+    assertEquals("123456", notification.getValue().getUserID(), "Error parsing userID");
+  }
+
+  @Test
+  public void onEvent_getCallFeedback() throws Exception {
+    final JsonNode message = mapper
+        .readTree(AsyncServiceTests.class.getResourceAsStream("/pubSubGetCallFeedback.json"));
+    final MvcResult mvcResult = mockMvc.perform(mockRequestBuilder
+        .content(message.toString())).andReturn();       
+    assertEquals(200, mvcResult.getResponse().getStatus());
+    ArgumentCaptor<TriggerEventNotification> notification =
+        ArgumentCaptor.forClass(TriggerEventNotification.class);
+    verify(asyncService).triggerEventHandler(notification.capture());
+    assertEquals(TriggerEventNotification.Event.GET_CALL_FEEDBACK,
+        notification.getValue().getEvent(), "Error parsing event");
+    assertEquals("+91 987456321",
+        notification.getValue().getEventParams().getFieldsMap()
+        .get("mobileNumber").getStringValue(),
+        "Error parsing userID");
+    assertEquals("123456", notification.getValue().getUserID(), "Error parsing userID");
+  }
+
+  @Test
+  public void onEvent_suggestImageUpload() throws Exception {
+    final JsonNode message = mapper
+        .readTree(AsyncServiceTests.class.getResourceAsStream("/pubSubSuggestImageUpload.json"));
+    final MvcResult mvcResult = mockMvc.perform(mockRequestBuilder
+        .content(message.toString())).andReturn();       
+    assertEquals(200, mvcResult.getResponse().getStatus());
+    ArgumentCaptor<TriggerEventNotification> notification =
+        ArgumentCaptor.forClass(TriggerEventNotification.class);
+    verify(asyncService).triggerEventHandler(notification.capture());
+    assertEquals(TriggerEventNotification.Event.SUGGEST_IMAGE_UPLOAD,
+        notification.getValue().getEvent(), "Error parsing event");
+    assertEquals("123456", notification.getValue().getUserID(), "Error parsing userID");
+  }
+
+  @Test(expected = Exception.class)
```

Comment:
Expect specific exception instead of Generic.
Same for other places

Reply:
updated